File: prototype/rails/src/test/domain/cart/cart_quantity_test.rb
require "test_helper"

module Domain
  module Cart
    class CartQuantityTest < ActiveSupport::TestCase
      test "it takes what was asked for when the stock covers it" do
        assert_equal 2, CartQuantity.within_stock(requested: 2, available: 3)
      end

      test "it stops at what is left" do
        assert_equal 3, CartQuantity.within_stock(requested: 5, available: 3)
      end

      test "it holds at least one of a listing" do
        assert_raises(ArgumentError) { CartQuantity.within_stock(requested: 0, available: 3) }
      end

      test "it refuses a sold out listing" do
        assert_raises(ArgumentError) { CartQuantity.within_stock(requested: 1, available: 0) }
      end
    end
  end
end

File: prototype/node/src/app/core/cart/cart-quantity.test.ts
import { test } from 'node:test'
import assert from 'node:assert/strict'
import { quantityWithinStock } from './cart-quantity.ts'

test('it takes what was asked for when the stock covers it', () => {
  assert.equal(quantityWithinStock({ requested: 2, available: 3 }), 2)
})

test('it stops at what is left', () => {
  assert.equal(quantityWithinStock({ requested: 5, available: 3 }), 3)
})

test('it holds at least one of a listing', () => {
  assert.throws(() => quantityWithinStock({ requested: 0, available: 3 }), RangeError)
})

test('it refuses a sold-out listing', () => {
  assert.throws(() => quantityWithinStock({ requested: 1, available: 0 }), RangeError)
})
